presto-main/src/main/java/com/facebook/presto/operator/HashAggregationOperator.java

```diff
@@ -323,6 +342,11 @@ public HashAggregationOperator(
         this.hashCollisionsCounter = new HashCollisionsCounter(operatorContext);
         operatorContext.setInfoSupplier(hashCollisionsCounter);
         this.useSystemMemory = useSystemMemory;
+        this.preGroupedHashStrategy = preGroupedChannels.isEmpty()
+                ? Optional.empty()
+                : Optional.of(joinCompiler.compilePagesHashStrategyFactory(
+                        preGroupedChannels.stream().map(channel -> groupByTypes.get(channel)).collect(toImmutableList()), preGroupedChannels, Optional.empty())
```

Comment:
nit: `groupByTypes::get`

Reply:
Checking if `preGroupedChannels` is fully contained in `groupByChannels` might not be necessary. It has been checked thoroughly during the [planning phase](https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddLocalExchanges.java#L334-L354).

Reply:
It would be really good to check the containment in the constructor. Planner and execution are usually separated. Just like Presto frontend and Velox backend. The modularization would break the streamlined assumption. Also, it would make it much clearer when we read the code to understand the logic.